app/models/internal/impl/PagerDutyNotificationEntry.java

```diff
@@ -0,0 +1,169 @@
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.steno.LoggerFactory;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.inject.Injector;
+import com.typesafe.config.Config;
+import models.internal.Alert;
+import models.internal.NotificationEntry;
+
+import java.net.URI;
+import java.util.Objects;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+/**
+ * Internal model representing a pagerduty notification entry.
+ * These do not have any internal state and act as singletons when creating a NotificationGroup.
+ *
+ * @author Sheldon White (sheldon.white at smartsheet dot com)
+ */
+public final class PagerDutyNotificationEntry implements NotificationEntry {
+    @Override
+    public CompletionStage<Void> notifyRecipient(final Alert alert, final AlertTrigger trigger, final Injector injector) {
+        LOGGER.debug().setMessage("executing pagerduty call").log();
+        final Config typesafeConfig = injector.getInstance(Config.class);
+        String pagerDutyEndpoint = typesafeConfig.getString("pagerDuty.uri");
+        String pagerDutyServiceKey = typesafeConfig.getString("pagerDuty.serviceKey");
+
+        try {
+            URI pagerDutyURI = new URI(pagerDutyEndpoint);
+
+            final ActorSystem actorSystem = injector.getInstance(ActorSystem.class);
+            final ObjectMapper mapper = injector.getInstance(ObjectMapper.class);
+            final Http http = Http.get(actorSystem);
+            final ObjectNode body = mapper.createObjectNode();
+            final String subject = String.format("Alert '%s' on %s in alarm ", alert.getName(), getGroupByString(trigger));
+            final String baseUrl = typesafeConfig.getString("alerts.baseUrl");
+            final String alertUrl = URI.create(baseUrl).resolve("/#alert/edit/" + alert.getId()).toString();
+
+            body.put("service_key", pagerDutyServiceKey);
+            body.put("event_type", "trigger");
+            body.put("description", subject);
+            body.set("alert", mapper.valueToTree(alert));
+            body.set("trigger", mapper.valueToTree(trigger));
+            body.put("alertUrl", alertUrl);
+            PagerDutyContext[] contexts = new PagerDutyContext[1];
+            PagerDutyContext context = new PagerDutyContext();
+            context.href = alertUrl;
+            context.type = "link";
+            context.text = "View the alert in M-Portal";
+            contexts[0] = context;
+            body.set("contexts", mapper.valueToTree(contexts));
+
+            return http
+                    .singleRequest(
+                            HttpRequest.POST(pagerDutyURI.toASCIIString())
+                                    .withEntity(HttpEntities.create(ContentTypes.APPLICATION_JSON, body.toString())))
+                    .thenApply(response -> {
+                        if (response.status().isFailure()) {
```

Comment:
When the POST to PagerDuty fails, is there a way to ensure retry?
Or, does the retry happen by virtue of the underlying alert being retriggered?

Reply:
That is up to the system.  It will certainly retry if the alert continues to trigger, but I believe it will also retry in cases of failure.  Either way, it shouldn't be handled here.